File: src/main/java/com/nao4j/subtrol/document/internal/Subscription.java
package com.nao4j.subtrol.document.internal;

import lombok.Data;
import lombok.NonNull;
import lombok.experimental.Wither;

import static com.nao4j.subtrol.document.internal.Quantity.QuantityType.UNITS;

@Data
@Wither
public class Subscription {

    @NonNull
    private final Price price;
    @NonNull
    private final RightOpenPeriod period;
    @NonNull
    private final Quantity quantity;

    public Subscription(final Price price, final RightOpenPeriod period, final Quantity quantity) {
        this.price = price;
        this.period = period;
        this.quantity = quantity;

        if (quantity.getType() == UNITS && period.getEnd() == null) {
            throw new IllegalArgumentException();
        }
    }
}

File: src/main/kotlin/com/nao4j/subtrol/model/Subscription.kt
package com.nao4j.subtrol.model

import com.nao4j.subtrol.model.Quantity.QuantityType.UNITS

data class Subscription(val price: Price, val period: RightOpenPeriod, val quantity: Quantity) {

    init {
        if (quantity.type == UNITS && period.end == null) {
            throw IllegalAccessException()
        }
        if (quantity.type == UNITS && period.end == null) {
            throw IllegalAccessException()
        }
    }

}
